tcmalloc/internal/percpu_rseq_aarch64.S

```diff
@@ -0,0 +1,492 @@
+// must match PERCPU_RSEQ_SIGNATURE (as configured by our rseq syscall's
+// signature parameter).  This signature is used to annotate valid abort IPs
+// (since rseq_cs could live in a user-writable segment).
+// We use .inst here instead of a data directive so it works for both small and
+// big endian.
+#define SIGN_ABORT()           \
+  .inst PERCPU_RSEQ_SIGNATURE
+
+/*
+ * Provide a directive to specify the size of symbol "label", relative to the
+ * current location and its start.
+ */
+#define ENCODE_SIZE(label) .size label, . - label
+/* We are assuming small memory model.  */
+#if !defined(__AARCH64_CMODEL_SMALL__)
+#error "Memory model not supported!"
+#endif
+
+/* FETCH_CPU assumes &__rseq_abi is in x5.  */
+#define FETCH_CPU(dest) \
+  ldr dest, [x5, #4] /* cpuid is 32-bits */
+
+/* With PIE  have initial-exec TLS, even in the presence of position
+   independent code. */
+#if !defined(__PIC__) || defined(__PIE__)
+
+#define START_RSEQ(src)				\
+  .L##src##_abort:				\
+  mrs	  x5, tpidr_el0;			\
+  adrp	  x6, :gottprel:__rseq_abi;		\
+  ldr	  x6, [x6,:gottprel_lo12:__rseq_abi];	\
+  add	  x5, x5, x6;				\
+  adrp    x6, __rseq_cs_##src;			\
+  add     x6, x6, :lo12:__rseq_cs_##src;	\
+  str     x6, [x5, #8];				\
+  .L##src##_start:
+
+#else  /* !defined(__PIC__) || defined(__PIE__) */
+
+/*
+ * In the case where we can't guarantee we have initial-exec TLS we use
+ * the General Dynamic TLS model to get the address of __rseq_abi.
+ * The call can be optimized away by the linker, but since we can not guarantee
+ * it will we must save and restore the registers used to store the arguments
+ * of our functions. The function with most arguments has 5 arguments, so we
+ * save x0-x4 and lr.
+ */
+#define START_RSEQ(src)				\
+  .L##src##_abort:				\
+  mov	  x5, lr;				\
+  stp	  x0, x1, [sp, -48]!;			\
+  stp	  x2, x3, [sp, #16];			\
+  stp	  x4, x5, [sp, #32];			\
+  adrp	  x0, :tlsdesc:__rseq_abi;		\
+  ldr	  x1, [x0, :tlsdesc_lo12:__rseq_abi];	\
+  add	  x0, x0, :tlsdesc_lo12:__rseq_abi;	\
+  .tlsdesccall __rseq_abi;			\
+  blr	  x1;					\
+  ldp	  x4, x5, [sp, #32];			\
+  mov     lr, x5;				\
+  mrs	  x5, tpidr_el0;			\
+  add	  x5, x5, x0;				\
+  ldp	  x2, x3, [sp, #16];			\
+  ldp	  x0, x1, [sp], #48;			\
+  adrp	  x6, __rseq_cs_##src;			\
+  add     x6, x6, :lo12:__rseq_cs_##src;	\
+  str     x6, [x5, #8];				\
+  .L##src##_start:
+
+#endif
+/* ---------------- end helper macros ---------------- */
+
+/* start of atomic restartable sequences */
+
+/*
+ * int TcmallocSlab_PerCpuCmpxchg64(int target_cpu, long *p,
+ *                                  long old_val, long new_val)
+ * w0: target_cpu
+ * x1: p
+ * x2: old_val
+ * x3: new_val
+ */
+  .p2align 6 /* aligns to 2^6 with NOP filling */
+  .globl TcmallocSlab_PerCpuCmpxchg64
+  .type  TcmallocSlab_PerCpuCmpxchg64, @function
+TcmallocSlab_PerCpuCmpxchg64:
+  .cfi_startproc
+.LTcmallocSlab_PerCpuCmpxchg64_region0:
```

Comment:
Nit: Since the region[0-3] labels aren't used, let's remove them.

(They appear in the x86_64 version mostly as a historical artifact of a previous implementation.)